### PR TITLE
feat: Integrate Gemini 1.5

### PR DESCRIPTION
**Pull request by Asher-hss**
## Description

I Integrate Gemini 1.5 API and add examples

## Motivation and Context

close https://github.com/camel-ai/camel/issues/557
close https://github.com/camel-ai/camel/issues/650

Integrate Google Deepmind's latest multi modal model Gemini 1.5

refer: https://ai.google.dev/gemini-api?hl=zh-cn

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [x ] Example (update in the folder of example)

## Implemented Tasks

- [ ] Subtask 1
- [ ] Subtask 2
- [ ] Subtask 3

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x ] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [ ] My change requires a change to the documentation.
- [x ] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: camel/agents/chat_agent.py [hunk 1]
```diff
@@ -448,6 +448,8 @@ def _step_model_response(
     ]:
         r"""Internal function for agent step model response."""
         # Obtain the model's response
+        print("-------------------------------------------")
+        print(openai_messages)
```
Comment (willshang76) [inline camel/agents/chat_agent.py:452]: Let's remove these debug prints.

File: camel/configs/gemini_config.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+                        `application/json`: JSON response in the candidates.
+
+                response_schema:
+                    Optional. Specifies the format of the JSON requested
+                    if response_mime_type is `application/json`.
+
+            safety_settings: Overrides for the model's safety settings.
+            stream: If True, yield response chunks as they are generated.
+            tools: `protos.Tools` more info coming soon.
+            request_options: Options for the request.
```
Comment (lightaime) [inline camel/configs/gemini_config.py:106]: The order of the docstring is not the same as the field. It is better to make them the same order

File: camel/models/gemini_model.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an “AS IS” BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+import os
+from typing import Any, Dict, Optional
+
+import google.generativeai as genai
```
Comment (lightaime) [inline camel/models/gemini_model.py:17]: Can we import inside `__init__` class?

File: camel/types/__init__.py [hunk 4]
```diff
@@ -26,6 +26,7 @@
     VectorDistance,
     VoiceType,
 )
+from .gemini_types import ContentsType, GenerateContentResponse
```
Comment (lightaime) [inline camel/types/__init__.py:29]: Not sure if we should implement it this way.

File: camel/models/gemini_model.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+            api_key (Optional[str]): The API key for authenticating with the
+                gemini service. (default: :obj:`None`)
+        """
+        super().__init__(model_type, model_config_dict, api_key)
+        self._api_key = api_key or os.environ.get("GOOGLE_API_KEY")
+        genai.configure(api_key=self._api_key)
+        if model_type.value not in ['gemini-1.5-flash', 'gemini-1.5-pro']:
+            raise ValueError(
+                "model_type: 'gemini-1.5-flash' or 'gemini-1.5-pro'"
+            )
```
Comment (lightaime) [inline camel/models/gemini_model.py:55]: We can remove this

**Review by lightaime — COMMENTED** [review]
Thanks @Asher-hss for the contribution! I left some comments. There are some improvements to be fix.

File: camel/configs/gemini_config.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+@dataclass(frozen=True)
+class Gemini_Gen_Config(BaseConfig):
+    candidate_count: Optional[int] = None
+    stop_sequences: Optional[Iterable[str]] = None
+    max_output_tokens: Optional[int] = None
+    temperature: Optional[float] = None
+    top_p: Optional[float] = None
+    top_k: Optional[int] = None
+    response_mime_type: Optional[str] = None
+    response_schema: Optional[protos.Schema] = None
```
[the remark this answers is not in the document]
Comment (Asher-hss) [reply camel/configs/gemini_config.py:134]: For Gemini, the parameters of Gemini_Gen_Config are not directly provided and need to be encapsulated into a fixed type, GenerationConfig. Gemini_Gen_Config provides its internal parameters.

[the remark this answers is not in the document]
Comment (Asher-hss) [reply camel/configs/gemini_config.py:134]: > What is the different between `GeminiConfig` and `Gemini_Gen_Config`?

For Gemini, the parameters of Gemini_Gen_Config are not directly provided and need to be encapsulated into a fixed type, GenerationConfig. Gemini_Gen_Config provides its internal parameters.

File: camel/configs/gemini_config.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+from camel.configs.base_config import BaseConfig
+
+
+@dataclass(frozen=True)
+class GeminiConfig(BaseConfig):
+    r"""A simple dataclass used to configure the generation parameters of
+    `GenerativeModel.generate_content`.
+
+    Arguments:
+            contents: The contents serving as the model's prompt.
```
[the remark this answers is not in the document]
Comment (Asher-hss) [reply camel/configs/gemini_config.py:36]: > It seems the `contents` is not a filed in the config.

For Gemini, content is equivalent to OpenAI's message.

File: camel/utils/token_counting.py [hunk 8]
```diff
@@ -297,10 +298,27 @@ def count_tokens_from_messages(self, messages: List[OpenAIMessage]) -> int:
...
+    def count_tokens_from_messages(self, messages) -> int:
+        r"""
+        Returns:
+        response(CountTokensResponse), use response.text to get output
+        """
+        if isinstance(messages, list):
+            for msg in messages:
+                if isinstance(msg, dict) and 'content' in msg:
+                    messages = msg['content']
+        return self._client.count_tokens(messages).total_tokens
```
[the remark this answers is not in the document]
Comment (Asher-hss) [reply camel/utils/token_counting.py:319]: > Can you explain?

The purpose of this part is essentially to create a converter that attempts to transform requests in the OpenAI format into requests in the Gemini format. I will adjust this part of the code in a more reasonable way.

File: camel/models/gemini_model.py [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+                role = 'user'
+            if 'content' in message:
+                converted_message = {
+                    "role": role,
+                    "parts": message.get("content"),
+                }
+                converted_messages.append(converted_message)
+            else:
+                converted_messages.append(message)
+        return converted_messages
```
Comment (Wendong-Fan) [inline camel/models/gemini_model.py:147]: docstring and type hint is missing

File: camel/models/gemini_model.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+                content = candidate.content.parts[0].text
+            choice = Choice(
+                index=i,
+                message=ChatCompletionMessage(
+                    role="assistant", content=content
+                ),
+                finish_reason='stop',
+            )
+        openai_res.choices.append(choice)
+        return openai_res
```
Comment (Wendong-Fan) [inline camel/models/gemini_model.py:172]: docstring and type hint is missing

File: camel/types/__init__.py
Comment (Wendong-Fan) [inline camel/types/__init__.py:1]: i think we don't need to include types from gemini in CAMEL

File: camel/types/gemini_types.py
Comment (Wendong-Fan) [inline camel/types/gemini_types.py:1]: I think it's not necessary to include gemini types in CAMEL

File: pyproject.toml [hunk 11]
```diff
@@ -39,6 +39,7 @@ anthropic = "^0.29.0"
 docstring-parser = "^0.15"
 pydantic = ">=1.9,<3"
 curl_cffi = "0.6.2"
+google-generativeai = "0.6.0"
```
Comment (Wendong-Fan) [inline pyproject.toml:42]: set google-generativeai as optional, you can refer to how litellm was set in toml file

**Review by Wendong-Fan — COMMENTED** [review]
Thanks for the contribution! Left some comments as below~

File: camel/configs/gemini_config.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    response_mime_type: Optional[str] = None
+    response_schema: Optional[protos.Schema] = None
+    safety_settings: Optional[safety_types.SafetySettingOptions] = None
+    stream: bool = False
+    tools: Optional[content_types.FunctionLibraryType] = None
+    tool_config: Optional[content_types.ToolConfigType] = None
+    request_options: Optional[helper_types.RequestOptionsType] = None
+
+    def to_dict(self):
+        return asdict(self)
```
[the remark this answers is not in the document]
Comment (Asher-hss) [reply camel/configs/gemini_config.py:129]: use this in role_playing example

File: camel/models/gemini_model.py [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+        you can directly output the response.text.
+
+        If it is in streaming mode,
+        you can iterate over the response chunks as they become available.
+        """
+        response = self._client.generate_content(
+            contents=self.to_gemini_req(contents),
+            **self.model_config_dict,
+        )
+        response.resolve()
```
Comment (Wendong-Fan) [inline camel/models/gemini_model.py:101]: could you help me understand why we need to run this? thanks!

File: camel/models/gemini_model.py [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+        return self.model_config_dict.get('stream', False)
+
+    def to_gemini_req(self, req):
+        r"""Converts the request from the OpenAI API format to
+            the Gemini API request format.
+
+        Args:
+            req: The request object from the OpenAI API.
+        Returns:
+            converted_messages: A list of messages formatted for Gemini API.
```
Comment (Wendong-Fan) [inline camel/models/gemini_model.py:136]:
```suggestion
        Args:
            req: The request object from the OpenAI API.

        Returns:
            converted_messages: A list of messages formatted for Gemini API.
```

File: camel/models/gemini_model.py [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+    @property
+    def stream(self) -> bool:
+        r"""Returns whether the model is in stream mode,
+            which sends partial results each time.
+        Returns:
+            bool: Whether the model is in stream mode.
+        """
+        return self.model_config_dict.get('stream', False)
+
+    def to_gemini_req(self, req):
```
Comment (Wendong-Fan) [inline camel/models/gemini_model.py:129]: add type hint for Args and response

File: camel/models/gemini_model.py [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+        Returns:
+            converted_messages: A list of messages formatted for Gemini API.
+        """
+        converted_messages = []
+        for message in req:
+            role = message.get('role')
+            if role == 'assistant':
+                role = 'model'
+            else:
+                role = 'user'
```
Comment (Wendong-Fan) [inline camel/models/gemini_model.py:144]: any reference for the convert here?

File: camel/models/gemini_model.py [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+
+        Returns:
+            response: A ChatCompletion object formatted for the OpenAI API.
+
+        If it is not in streaming mode,
+        you can directly output the response.text.
+
+        If it is in streaming mode,
+        you can iterate over the response chunks as they become available.
+        """
```
Comment (Wendong-Fan) [inline camel/models/gemini_model.py:96]: if it's in stream mode, we may also need to convert the response into OpenAI format

File: camel/models/gemini_model.py [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+                        f"Unexpected argument `{param}` is "
+                        "input into Gemini model backend."
+                    )
+
+    @property
+    def stream(self) -> bool:
+        r"""Returns whether the model is in stream mode,
+            which sends partial results each time.
+        Returns:
+            bool: Whether the model is in stream mode.
```
Comment (Wendong-Fan) [inline camel/models/gemini_model.py:125]:
```suggestion
        r"""Returns whether the model is in stream mode, which sends partial
        results each time.

        Returns:
            bool: Whether the model is in stream mode.
```

File: camel/models/gemini_model.py [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+                converted_message = {
+                    "role": role,
+                    "parts": message.get("content"),
+                }
+                converted_messages.append(converted_message)
+            else:
+                converted_messages.append(message)
+        return converted_messages
+
+    def to_openai_response(self, res):
```
Comment (Wendong-Fan) [inline camel/models/gemini_model.py:155]: type hint

File: camel/models/gemini_model.py [hunk 20]
```diff
@@ -0,0 +1,185 @@
...
+        r"""Converts the response from the Gemini API to the OpenAI API
+        response format.
+        Args:
+            res: The response object returned by the Gemini API
+        Returns:
+            openai_res: A ChatCompletion object formatted for the OpenAI API.
+        """
```
Comment (Wendong-Fan) [inline camel/models/gemini_model.py:162]:
```suggestion
        r"""Converts the response from the Gemini API to the OpenAI API
        response format.

        Args:
            res: The response object returned by the Gemini API.

        Returns:
            openai_res: A ChatCompletion object formatted for the OpenAI API.
        """
```

File: camel/models/gemini_model.py [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+        for i, candidate in enumerate(res.candidates):
+            content = ""
+            if candidate.content and len(candidate.content.parts) > 0:
+                content = candidate.content.parts[0].text
+            choice = Choice(
+                index=i,
+                message=ChatCompletionMessage(
+                    role="assistant", content=content
+                ),
+                finish_reason='stop',
```
Comment (Wendong-Fan) [inline camel/models/gemini_model.py:182]: it's may not good to hard code the role and finish reason, if these information is not provided from gemini, maybe we can just set this as None

File: camel/utils/token_counting.py [hunk 22]
```diff
@@ -303,6 +303,37 @@ def count_tokens_from_messages(self, messages: List[OpenAIMessage]) -> int:
...
 
+class GeminiTokenCounter(BaseTokenCounter):
+    def __init__(self, model_type: ModelType):
+        r"""Constructor for the token counter for Gemini models."""
+        import google.generativeai as genai
+
+        self.model_type = model_type
+        self._client = genai.GenerativeModel(self.model_type.value)
+
+    def count_tokens_from_messages(self, messages) -> int:
```
Comment (Wendong-Fan) [inline camel/utils/token_counting.py:314]: type hint

File: camel/utils/token_counting.py [hunk 23]
```diff
@@ -303,6 +303,37 @@ def count_tokens_from_messages(self, messages: List[OpenAIMessage]) -> int:
...
+        r"""
+        Returns:
+        response(CountTokensResponse), use response.text to get output
+        """
```
Comment (Wendong-Fan) [inline camel/utils/token_counting.py:318]:
```suggestion
        r"""Count number of tokens in the provided message list using
        loaded tokenizer specific for this type of model.

        Args:
            messages (List[OpenAIMessage]): Message list with the chat history
                in OpenAI API format.

        Returns:
            int: Number of tokens in the messages.
        """
```

File: examples/models/role_playing_with_gemini.py [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+
+
+def main(model_type=None) -> None:
+    task_prompt = "Develop a trading bot for the stock market"
+
+    agent_kwargs = {
+        role: ModelFactory.create(
+            model_platform=ModelPlatformType.GEMINI,
+            model_type=model_type,
+            model_config_dict=dataclasses.asdict(GeminiConfig()),
```
Comment (Wendong-Fan) [inline examples/models/role_playing_with_gemini.py:32]:
```suggestion
            model_config_dict=GeminiConfig().__dict__,
```

File: pyproject.toml [hunk 25]
```diff
@@ -43,6 +43,7 @@ pillow ="^10.2.0"
 
 # model-platforms
 litellm = { version = "^1.38.1", optional = true }
+google-generativeai = { version = "0.6.0", optional = true }
```
Comment (Wendong-Fan) [inline pyproject.toml:46]:
```suggestion
google-generativeai = { version = "^0.6.0", optional = true }
```

**Review by Wendong-Fan — COMMENTED** [review]
Thanks @Asher-hss's update and sorry for the late review, left some comments

File: camel/models/gemini_model.py (hunk 13 above)
Comment (Asher-hss) [reply camel/models/gemini_model.py:101]: if it is stream mode, we need to use this function to deal with the response of Gemini model

**Comment by Wendong-Fan** [comment]
Hey @Asher-hss , you need to update your master branch and resolve the conflicts, let me know if you need further support on this~

**Review by Wendong-Fan — APPROVED** [review]
Thanks @Asher-hss 's contribution!

**Pull request merged**